pitop/robotics/navigation/navigation_controller.py

```diff
@@ -0,0 +1,325 @@
+from threading import Thread, Event
+from time import time
+from dataclasses import dataclass
+import numpy as np
+import math
+from typing import Union
+from simple_pid import PID
+
+
+@dataclass
+class RobotState:
+    x: float = 0.0
+    y: float = 0.0
+    theta: float = 0.0
+    v: float = 0.0
+    w: float = 0.0
+
+    @property
+    def pose(self):
+        return self.x, self.y, self.theta
+
+    def reset_pose(self):
+        self.x, self.y, self.theta = (0.0, 0.0, 0.0)
+
+    def calc_distance(self, point_x, point_y):
+        dx = self.x - point_x
+        dy = self.y - point_y
+        return math.hypot(dx, dy)
+
+
+class GoalCriteria:
+    def __init__(self, linear_speed_factor, angular_speed_factor):
+        self._MAX_GOAL_REACHED_DISTANCE_ERROR = linear_speed_factor * 0.02  # 2 mm at full speed
+        self._MAX_GOAL_REACHED_ANGLE_ERROR = angular_speed_factor * math.radians(4.0)  # 4deg at full speed
+
+    def angle(self, angle_error):
+        if abs(angle_error) < self._MAX_GOAL_REACHED_ANGLE_ERROR:
+            return True
+        return False
+
+    def distance(self, distance_error, heading_error):
+        if abs(distance_error) < self._MAX_GOAL_REACHED_DISTANCE_ERROR:
+            return True
+        if abs(heading_error) > math.pi / 2:
+            # Overshot goal, error will be small so consider goal to be reached.
+            # Otherwise robot would have to reverse a tiny amount which would be poor UX.
+            # Robot will correct itself on subsequent navigation calls anyway.
+            return True
+        return False
+
+
+class RobotDrivingParameters:
+    def __init__(self, max_motor_speed, max_angular_speed):
+        self._max_motor_speed = max_motor_speed
+        self._max_angular_speed = max_angular_speed
+
+        self.linear_speed_factor = 0.0
+        self.angular_speed_factor = 0.0
+        self.max_v = 0.0
+        self.max_w = 0.0
+        self.deceleration_distance = 0.0
+        self.deceleration_angle = 0.0
+
+    def update_linear_speed(self, speed_factor):
+        self.linear_speed_factor = speed_factor
+        self.max_v = self.linear_speed_factor * self._max_motor_speed
+        self.deceleration_distance = self.linear_speed_factor * 0.4  # 0.4m at full speed
```

Comment:
I think we could set this `0.4` and `120` at the `update_angular_speed` method as class attributes